swim/state_transitions.go

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package swim
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	log "github.com/uber-common/bark"
+	"github.com/uber/ringpop-go/logging"
+	"github.com/uber/ringpop-go/util"
+)
+
+// subject is an interface to define the subject (eg. member) to transition state for
+type subject interface {
+	address() string
+	incarnation() int64
+}
+
+type transitionTimer struct {
+	*time.Timer
+
+	// state represents the state the subject was in when the transition was scheduled
+	state string
+
+	// transition contains the transition to execute
+	transition func()
+}
+
+// stateTransitions handles the timers for state transitions in SWIM
+type stateTransitions struct {
+	sync.Mutex
+
+	node *Node
+
+	timeouts StateTimeouts
+	timers   map[string]*transitionTimer
+	enabled  bool
+	logger   log.Logger
+}
+
+// StateTimeouts contains the timeouts for every state before transitioning to the new state
+type StateTimeouts struct {
+	// Suspect is the timeout it takes a node in suspect mode to transition to faulty
+	Suspect time.Duration
+}
+
+// GetTimeout returns the timeout to apply on a transition from state
+func (s *StateTimeouts) GetTimeout(state string) time.Duration {
+	switch state {
+	case Suspect:
+		return s.Suspect
+	default:
+		panic(fmt.Sprintf("unknown state %q", state))
+	}
+}
```

Comment:
We can get rid of the `switch` here if we pass the timer explicitly in `ScheduleSuspectToFaulty`.